WordPress/Sniffs/WP/I18nSniff.php

```diff
@@ -0,0 +1,256 @@
+			if ( ! empty( $this_token['parenthesis_closer'] ) ) {
+				for ( $j = $i + 1; $j <= $this_token['parenthesis_closer']; $j += 1 ) {
+					$tokens[ $j ]['token_index'] = $j;
+					$argument_tokens[] = $tokens[ $j ];
+				}
+				$i = $this_token['parenthesis_closer'];
+			}
+		}
+		if ( ! empty( $argument_tokens ) ) {
+			$arguments_tokens[] = $argument_tokens;
+		}
+		unset( $argument_tokens );
+
+		$argument_assertions = array();
+		if ( in_array( $translation_function, array( '__', 'esc_attr__', 'esc_html__', '_e', 'esc_attr_e', 'esc_html_e', 'translate' ) ) ) {
+			$argument_assertions[] = array( '$text', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		} else if ( in_array( $translation_function, array( '_x', '_ex', 'esc_attr_x', 'esc_html_x', 'translate_with_gettext_context' ) ) ) {
+			$argument_assertions[] = array( '$text', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$context', 'check_literal_string_context_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		} else if ( in_array( $translation_function, array( '_n', '_n_noop' ) ) ) {
+			$argument_assertions[] = array( '$single', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$plural', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$number', 'check_number_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		} else if ( in_array( $translation_function, array( '_nx', '_nx_noop' ) ) ) {
+			$argument_assertions[] = array( '$single', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$plural', 'check_literal_string_text_tokens' );
+			$argument_assertions[] = array( '$number', 'check_number_tokens' );
+			$argument_assertions[] = array( '$context', 'check_literal_string_context_tokens' );
+			$argument_assertions[] = array( '$domain', 'check_string_domain_tokens' );
+		}
+
+		$argument_stack_ptr = $func_open_paren_token;
+		foreach ( $argument_assertions as $argument_assertion ) {
+			$argument_tokens = array_shift( $arguments_tokens );
+			if ( ! $argument_tokens ) {
+				$argument_tokens = array();
+			} else {
+				$argument_stack_ptr = $argument_tokens[0]['token_index'];
+			}
+			$method_name = $argument_assertion[1];
+			$arg_name = $argument_assertion[0];
+			call_user_func( array( $this, $method_name ), $phpcs_file, $arg_name, $argument_stack_ptr, $argument_tokens );
+		}
+	}
+
+	/**
+	 * Check if supplied tokens represent a translation text string literal.
+	 *
+	 * @param PHP_CodeSniffer_File $phpcs_file The file being scanned.
+	 * @param string $arg_name
+	 * @param int $stack_ptr
+	 * @param array $tokens
+	 * @return bool
+	 */
+	protected function check_literal_string_text_tokens( $phpcs_file, $arg_name = '$text', $stack_ptr, $tokens = array() ) {
```

Comment:
Also, some are errors and some are warnings.

It could be refactored to try to abstract these out more by passing in params to common methods if you think it is worthwhile.